src/Umbraco.Web.Common/AspNetCore/AspNetCoreHostingEnvironment.cs

```diff
@@ -37,7 +37,16 @@ public class AspNetCoreHostingEnvironment : IHostingEnvironment
             _webHostEnvironment = webHostEnvironment ?? throw new ArgumentNullException(nameof(webHostEnvironment));
             _urlProviderMode = _webRoutingSettings.CurrentValue.UrlProviderMode;
 
-            SiteName = webHostEnvironment.ApplicationName;
+            SetSiteName(hostingSettings.CurrentValue.SiteName);
+
+            // We have to ensure that the OptionsMonitor is an actual options monitor since we have a hack
+            // where we initially use an OptionsMonitorAdapter, which doesn't implement OnChange.
+            // See summery of OptionsMonitorAdapter for more information.
+            if (hostingSettings is OptionsMonitor<HostingSettings>)
+            {
+                hostingSettings.OnChange(settings => SetSiteName(settings.SiteName));
+            }
```

Comment:
Why not make the `OptionsMonitorAdapter.OnChange()` a no-op/not throw the `NotImplementedException` instead?

Reply:
Seems reasonable enough to me